File: packages/react-native-enriched-markdown/android/src/main/java/com/swmansion/enriched/markdown/input/formatting/RangeEditAdjustment.kt
package com.swmansion.enriched.markdown.input.formatting

import com.swmansion.enriched.markdown.input.model.MutableRangeBounds

internal fun <T : MutableRangeBounds> sortedInsertionIndex(
  ranges: List<T>,
  location: Int,
): Int {
  var index = 0
  for (existing in ranges) {
    if (existing.start > location) break
    index++
  }
  return index
}

/**
 * Shared shift/clip logic applied to stored ranges after a text edit. Both
 * [FormattingStore] and [BlockStore] delegate here so the overlap
 * classification lives in exactly one place.
 */
internal object RangeEditAdjustment {
  /**
   * Mutates [ranges] in place to follow an edit that replaced [deletedLength]
   * characters at [editLocation] with [insertedLength] characters. Ranges
   * deleted outright or clipped to zero length are removed.
   *
   * Insert-only edits at exactly `range.start` or `range.end` do NOT grow the
   * range — the typed characters stay outside it. Whether boundary text joins
   * the range is decided elsewhere: pending styles for inline ranges, line
   * re-normalization for block ranges.
   *
   * [inheritsReplacementAtStart]: when it returns true for a range whose start
   * is the edit location, replacement text joins the range (UIKit attribute
   * inheritance); when false, the old clip/remove behavior applies.
   */
  fun <T : MutableRangeBounds> adjustForEdit(
    ranges: MutableList<T>,
    editLocation: Int,
    deletedLength: Int,
    insertedLength: Int,
    inheritsReplacementAtStart: (T) -> Boolean = { false },
  ) {
    if (deletedLength == 0 && insertedLength == 0) return

    val deleteEnd = editLocation + deletedLength
    val indexesToRemove = mutableListOf<Int>()

    for ((idx, range) in ranges.withIndex()) {
      if (deletedLength > 0) {
        val inheritsReplacement =
          insertedLength > 0 && range.start == editLocation && inheritsReplacementAtStart(range)
        when (classifyOverlap(range.start, range.end, editLocation, deleteEnd)) {
          EditOverlap.BEFORE_EDIT -> { /* no change */ }

          EditOverlap.AFTER_EDIT -> {
            range.start = range.start - deletedLength + insertedLength
            range.end = range.end - deletedLength + insertedLength
          }

          EditOverlap.FULLY_DELETED -> {
            if (inheritsReplacement) {
              range.start = editLocation
              range.end = editLocation + insertedLength
            } else {
              indexesToRemove.add(idx)
            }
          }

          EditOverlap.DELETED_INSIDE -> {
            range.end = range.end - deletedLength + insertedLength
          }

          EditOverlap.CLIPPED_END -> {
            val newEnd = editLocation + insertedLength
            val newLength = if (newEnd > range.start) newEnd - range.start else 0
            range.end = range.start + newLength
            if (newLength == 0) indexesToRemove.add(idx)
          }

          EditOverlap.CLIPPED_START -> {
            val charsClipped = deleteEnd - range.start
            val survivingLength = range.end - range.start - charsClipped
            if (inheritsReplacement) {
              range.start = editLocation
              range.end = editLocation + insertedLength + survivingLength
            } else {
              range.start = editLocation + insertedLength
              range.end = range.start + survivingLength
              if (survivingLength == 0) indexesToRemove.add(idx)
            }
          }
        }
      } else {
        when {
          range.start >= editLocation -> {
            range.start += insertedLength
            range.end += insertedLength
          }

          editLocation > range.start && editLocation < range.end -> {
            range.end += insertedLength
          }
        }
      }
    }

    for (idx in indexesToRemove.reversed()) {
      ranges.removeAt(idx)
    }

    ranges.removeAll { it.end - it.start == 0 }
  }

  private enum class EditOverlap {
    BEFORE_EDIT,
    AFTER_EDIT,
    FULLY_DELETED,
    DELETED_INSIDE,
    CLIPPED_END,
    CLIPPED_START,
  }

  private fun classifyOverlap(
    rangeStart: Int,
    rangeEnd: Int,
    editLocation: Int,
    deleteEnd: Int,
  ): EditOverlap {
    if (rangeEnd <= editLocation) return EditOverlap.BEFORE_EDIT
    if (rangeStart >= deleteEnd) return EditOverlap.AFTER_EDIT
    if (rangeStart >= editLocation && rangeEnd <= deleteEnd) return EditOverlap.FULLY_DELETED
    if (rangeStart < editLocation && rangeEnd > deleteEnd) return EditOverlap.DELETED_INSIDE
    return if (rangeStart < editLocation) EditOverlap.CLIPPED_END else EditOverlap.CLIPPED_START
  }
}

File: packages/react-native-enriched-markdown/src/web/input/formatting/RangeEditAdjustment.ts
import type { RangeBounds } from '../model/rangeBounds';

type EditOverlap =
  | 'before-edit'
  | 'after-edit'
  | 'fully-deleted'
  | 'deleted-inside'
  | 'clipped-end'
  | 'clipped-start';

function classifyOverlap(
  rangeStart: number,
  rangeEnd: number,
  editLocation: number,
  deleteEnd: number
): EditOverlap {
  if (rangeEnd <= editLocation) return 'before-edit';
  if (rangeStart >= deleteEnd) return 'after-edit';
  if (rangeStart >= editLocation && rangeEnd <= deleteEnd) {
    return 'fully-deleted';
  }
  if (rangeStart < editLocation && rangeEnd > deleteEnd) {
    return 'deleted-inside';
  }
  return rangeStart < editLocation ? 'clipped-end' : 'clipped-start';
}

/**
 * Shift/clip logic applied to stored ranges after a text edit that replaced
 * `deletedLength` characters at `editLocation` with `insertedLength`
 * characters. Bounds are mutated in place; the returned array drops ranges
 * deleted outright or clipped to zero length.
 *
 * Insert-only edits at exactly `range.end` do NOT grow the range — whether
 * typed text continues a style is decided by the pending-styles layer, and
 * links must never auto-extend. An insert at
 * exactly `range.start` grows the range only when `growsAtStartOnInsert`
 * returns true (block ranges own their whole line); otherwise the range shifts
 * and the typed characters stay outside it (a character typed before a bold
 * run must not become bold).
 *
 * `inheritsReplacementAtStart`: when true for a range whose start is the edit
 * location, replacement text joins the range (autocorrect over a styled word
 * keeps the style); when false, the plain clip/remove behavior applies.
 */
export function adjustRangesForEdit<T extends RangeBounds>(
  ranges: T[],
  editLocation: number,
  deletedLength: number,
  insertedLength: number,
  inheritsReplacementAtStart: (range: T) => boolean = () => false,
  growsAtStartOnInsert: (range: T) => boolean = () => false
): T[] {
  if (deletedLength === 0 && insertedLength === 0) {
    return ranges;
  }

  const deleteEnd = editLocation + deletedLength;
  const removed = new Set<T>();

  for (const range of ranges) {
    if (deletedLength > 0) {
      const inheritsReplacement =
        insertedLength > 0 &&
        range.start === editLocation &&
        inheritsReplacementAtStart(range);

      switch (
        classifyOverlap(range.start, range.end, editLocation, deleteEnd)
      ) {
        case 'before-edit':
          break;

        case 'after-edit':
          range.start += insertedLength - deletedLength;
          range.end += insertedLength - deletedLength;
          break;

        case 'fully-deleted':
          if (inheritsReplacement) {
            range.start = editLocation;
            range.end = editLocation + insertedLength;
          } else {
            removed.add(range);
          }
          break;

        case 'deleted-inside':
          range.end += insertedLength - deletedLength;
          break;

        case 'clipped-end': {
          const newEnd = editLocation + insertedLength;
          const newLength = newEnd > range.start ? newEnd - range.start : 0;
          range.end = range.start + newLength;
          if (newLength === 0) {
            removed.add(range);
          }
          break;
        }

        case 'clipped-start': {
          const charsClipped = deleteEnd - range.start;
          const survivingLength = range.end - range.start - charsClipped;
          if (inheritsReplacement) {
            range.start = editLocation;
            range.end = editLocation + insertedLength + survivingLength;
          } else {
            range.start = editLocation + insertedLength;
            range.end = range.start + survivingLength;
            if (survivingLength === 0) {
              removed.add(range);
            }
          }
          break;
        }
      }
    } else {
      if (range.start === editLocation && growsAtStartOnInsert(range)) {
        range.end += insertedLength;
      } else if (range.start >= editLocation) {
        range.start += insertedLength;
        range.end += insertedLength;
      } else if (editLocation > range.start && editLocation < range.end) {
        range.end += insertedLength;
      }
    }
  }

  return ranges.filter(
    (range) => !removed.has(range) && range.end - range.start > 0
  );
}
